### ISSUE TITLE
vault 1.9 agent cannot use address

### ISSUE THREAD

**Issue by tingShean** [hashicorp/vault#13363]
**Describe the bug**
i use vault agent for render secret data, but try some days it's just only use 127.0.0.1, any enviroment cannot override this config

**this bug in vault version 1.9, i change to 1.8.4 is worked**

**To Reproduce**
1. Run `vault agent -config /etc/vault.d/`
2. Run `systemctl start vault`

will see this error and get like ddos attack your server to CPU 100%
```
2021-12-08T07:12:16.719Z [ERROR] template.server: template server error:
  error=
  | vault.read(secret/my-secret/data): vault.read(secret/my-secret/data): Error making API request.
  | 
  | URL: GET http://127.0.0.1:8200/v1/secret/my-secret/data
  | Code: 403. Errors:
  | 
  | * permission denied
```

**Expected behavior**
it will use config file's vault address, not localhost

**Environment:**
* Vault Server Version (retrieve with `vault status`): 1.8.4 (925bc650ad1d997e84fbb832f302a6bfe0105bbb)
* Vault Agent Version : 1.9.0
* Vault CLI Version (retrieve with `vault version`): 1.9.0
* Server Operating System/Architecture: Centos 7

**Comment by hsimon-hashicorp** [comment]
Hi there! Can you show your Vault Agent config, too, with any sensitive information redacted?

**Comment by tingShean** [comment]
this agent config
```hcl
pid_file = "/etc/vault.d/pidfile"
                                                                                                                                                                                                                          
vault {
  address = "https://mydomain.com"
}

auto_auth {

  method {

    type      = "approle"

    config = {
      role_id_file_path = "/etc/vault.d/roleid"
      secret_id_file_path = "/etc/vault.d/.secretid"
      # for test open
      remove_secret_id_file_after_reading = false
    }
  }

  sink "file" {
    config = {
      path = "/etc/vault.d/vault-token-via-agent"
    }
  }
}

storage "raft" {
  path    = "/var/vault/data"
  node_id = "vault-02"

  retry_join {
    leader_api_addr = "https://mydomain.com"
  }
}

cache {
  use_auto_auth_token = true
}

listener "tcp" {
  address = "127.0.0.1:8100"
  tls_disable = true
}

template {
  source      = "/etc/vault.d/my.tmpl"
  destination = "/home/user/my.secret"
  perms       = "0600"
}
```

agent get mydomain.com in init, but nevert use

use yum install vault default version is 1.9.0 now

i have other agent working fine in version 1.8.4 and use yum install before

remove the version and install specify 1.8.4 is worked

**Comment by hsimon-hashicorp** [comment]
Can you change the listener stanza to specify your preferred address and see if that works? Please see the following: https://www.vaultproject.io/docs/agent#listener

Thanks!

**Comment by tingShean** [comment]
i try version 1.9.1 on new server

if add listener vault master address, can not start vault agent

and change to 127.0.0.1:8200 or remove, got the same result that like DDoS after start agent few second, log always Error 403

this configure is worked on version 1.8.4 on other server, if remove template and listener, error message say start agent  need template or listener.

don't know why cannot run start agent when i just remove listener.

**Comment by pmmukh** [comment]
Hi @tingShean ! I was looking into this issue today, and have been able to reproduce the error you were seeing. However, I don't think it is because it was unable to talk to the Vault server. The 403 permission denied error would not be possible, if it was a connectivity issue, in which case I would expect a 5XX error response. Rather, I suspect it is an issue with the error log, that prints the vault address incorrectly.
In my case, the issue was not an error parsing the address from the config file, but rather that my approle token did not have permissions to access the secret path I had defined in my template file. I suspect you may be having the same issue, based on the error code, could you double check the policies on your app role once?
There are 2 things at play here that I can see:
1. When the template server gets back an error response from the Vault server, when logging the error, it incorrectly prints the localhost address instead of the address actually in use. I will need to dive into the code to verify this, but I think that's what is at play here. I think this should be solved as part of this issue.
2. The DoS issue of the template server, when an api call to Vault fails it should engage in some sort of exponential backoff instead of retrying and failing in a tight loop. (This happened to me as well) This has already been called out here https://github.com/hashicorp/vault/issues/12566 and should be addressed in the fix for that issue.

**Comment by tingShean** [comment]
Hi @pmmukh 

I use the same configure on different version server(1.8.4, 1.9.0 and 1.9.1) and it is worked only version 1.8.4. 

the specified vault server address just show once when startup, then show connect 127.0.0.1 not stop if you do not stop it or kill the process. 

i reference official document for template config, this my [config](https://github.com/hashicorp/vault/issues/13363#issuecomment-989448045)

**Comment by pmmukh** [comment]
Hey @tingShean ! Yeah your config file looks fine to me, that I don't think is the issue, although in my case I did use the port also, not exactly sure what the default port is when not specified, for one thing could you try adding the port into the vault address you're setting in the config ? 
I think this is getting confusing because I don't have a clear picture of what is being done on the vault side i.e how the secret and the approle is being setup, and how the template file is setup.
Could you please add a list of vault cli commands to this issue, that shows your vault side setup, your template file setup, and then the command to start your vault agent ?

**Comment by dekimsey** [comment]
@pmmukh For what it's worth, I'm seeing this issue as well. Interestingly my logs show the agent trying to hit localhost and the configured vault addr later on. Though it's dying for reasons related to the DOS issue in templating. I have a full log excerpt available in #13589 if that's helpful.

**Comment by VioletHynes** [comment]
Hi there! The DOS issue  this will be fixed by https://github.com/hashicorp/vault/pull/16970 - if you have caching in the Vault Agent config, currently, Agent will ignore the retry configuration and perform immediate infinite retries in a loop with no back-off. As a workaround, an empty `template_config` stanza should work, by setting retry to the default config (12), though obviously that's not perfect as you might not want 12.

The fix should be coming in 1.12.

**Comment by VioletHynes** [comment]
Hi there! I'm going to close this issue as I just merged https://github.com/hashicorp/vault/pull/16970 which should fix the DOS issue identified here. This should release in 1.12.

The localhost you're seeing should be explained by this comment: https://github.com/hashicorp/vault/issues/13589#issuecomment-1007028847

Thanks for the bug report!

Do feel free to reopen if I've missed something or this doesn't solve your issues.
